nav2_amcl/src/amcl_node.cpp

```diff
@@ -20,440 +20,219 @@
-  init_cov_[1] = 0.5 * 0.5;
-  init_cov_[2] = (M_PI / 12.0) * (M_PI / 12.0);
-
-  // TODO(mhpanah): Enable reading pose from parameter server.
-/*
-  // Check for NAN on input from param server, #5239
-  double tmp_pos;
+  rclcpp::Time last_error = rclcpp_node_->now();
+  std::string tf_error;
 
-  tmp_pos = parameters_client->get_parameter("initial_pose_x", init_pose_[0]);
+  RCLCPP_INFO(get_logger(), "Checking that transform thread is ready");
 
-  if (!std::isnan(tmp_pos)) {
-    init_pose_[0] = tmp_pos;
-  } else {
-    RCLCPP_WARN(get_logger(), "ignoring NAN in initial pose X position");
-  }
-
-  tmp_pos = parameters_client->get_parameter("initial_pose_y", init_pose_[1]);
-
-  if (!std::isnan(tmp_pos)) {
-    init_pose_[1] = tmp_pos;
-  } else {
-    RCLCPP_WARN(get_logger(), "ignoring NAN in initial pose Y position");
-  }
+  while (rclcpp::ok() &&
+    !tf_buffer_->canTransform(global_frame_id_, odom_frame_id_, tf2::TimePointZero,
+    tf2::durationFromSec(0.1), &tf_error))
+  {
+    if (last_error + nav2_util::duration_from_seconds(1.0) < rclcpp_node_->now()) {
+      RCLCPP_INFO(get_logger(), "Timed out waiting for transform from %s to %s"
+        " to become available, tf error: %s",
+        odom_frame_id_.c_str(), global_frame_id_.c_str(), tf_error.c_str());
+      last_error = rclcpp_node_->now();
+    }
 
-  tmp_pos = parameters_client->get_parameter("initial_pose_a", init_pose_[2]);
+    // The error string will accumulate and errors will typically be the same, so the last
+    // will do for the warning above. Reset the string here to avoid accumulation.
+    tf_error.clear();
 
-  if (!std::isnan(tmp_pos)) {
-    init_pose_[2] = tmp_pos;
-  } else {
-    RCLCPP_WARN(get_logger(), "ignoring NAN in initial pose Yaw");
+    // Let the transforms populate, then retry
+    std::this_thread::sleep_for(10ms);
   }
+}
 
-  tmp_pos = parameters_client->get_parameter("initial_cov_xx", init_cov_[0]);
+nav2_lifecycle::CallbackReturn
+AmclNode::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Activating");
 
-  if (!std::isnan(tmp_pos)) {
-    init_cov_[0] = tmp_pos;
-  } else {
-    RCLCPP_WARN(get_logger(), "ignoring NAN in initial covariance XX");
-  }
+  // Lifecycle publishers must be explicitly activated
+  pose_pub_->on_activate();
+  particlecloud_pub_->on_activate();
 
-  tmp_pos = parameters_client->get_parameter("initial_cov_yy", init_cov_[1]);
+  first_pose_sent_ = false;
 
-  if (!std::isnan(tmp_pos)) {
-    init_cov_[1] = tmp_pos;
-  } else {
-    RCLCPP_WARN(get_logger(), "ignoring NAN in initial covariance YY");
-  }
+  // Keep track of whether we're in the active state. We won't
+  // process incoming callbacks until we are
+  active_ = true;
 
-  tmp_pos = parameters_client->get_parameter("initial_cov_aa", init_cov_[2]);
+  // Wait until the transform listener thread has had a chance to spin up and get the
+  // transforms that it needs
+  waitForTransforms();
```

Comment:
@crdelsey Reliable, yes, but not necessarily quick. IMO we want to make sure we've got what we need before continuing. If not, something's wrong. I have purposely not added timeouts yet so that I know exactly where there is a problem and can fix it. Except for the occasional gazebo crashes, the system comes up fairly reliably. I'd like to continue with this approach and then discuss timeouts, system monitoring, error recovery, etc. with a system perspective in mind.